### PR TITLE
Create callback trait for peer set changes.

### PR DESCRIPTION
**Pull request by arik-so**

### CODE REVIEW

File: lightning/src/ln/peer_handler.rs [hunk 1]
```diff
@@ -236,6 +236,14 @@ impl error::Error for PeerHandleError {
 	}
 }
 
+/// Trait for subscribing to changes of a peer manager's peer set
+pub trait PeerConnectivityUpdater {
```
Comment (jkczyz) [inline lightning/src/ln/peer_handler.rs:240]: For naming maybe simply `Connectivity`. Alternative: `Listen` similar to listening for blocks though might cause some problems in bindings if the name is reused.

File: lightning/src/ln/peer_handler.rs [hunk 2]
```diff
@@ -354,7 +362,11 @@ pub struct PeerManager<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: De
 	peer_counter_low: AtomicUsize,
 	peer_counter_high: AtomicUsize,
 
+	// Optional, if the user wishes to be notified about changes to the peer set
+	peer_connectivity_updater: Option<Box<dyn PeerConnectivityUpdater>>,
```
Comment (jkczyz) [inline lightning/src/ln/peer_handler.rs:366]: For consistency with our codebase, should we use a template parameter? May also help resolve the compilation error:

```term
error: future cannot be sent between threads safely
   --> lightning-net-tokio/src/lib.rs:274:8
    |
274 |         Some(tokio::spawn(async move {
    |              ^^^^^^^^^^^^ future created by async block is not `Send`
    | 
   ::: /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.8.1/src/task/spawn.rs:127:21
    |
127 |         T: Future + Send + 'static,
    |                     ---- required by this bound in `tokio::spawn`
    |
    = help: the trait `Send` is not implemented for `(dyn PeerConnectivityUpdater + 'static)`
note: captured value is not `Send`
   --> lightning-net-tokio/src/lib.rs:292:8
    |
292 | ...                   peer_manager.socket_disconnected(&SocketDescriptor::new(Arc::clone(&us)));
    |                       ^^^^^^^^^^^^ has type `Arc<PeerManager<SocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>>>` which is not `Send`
```

File: lightning/src/ln/peer_handler.rs [hunk 3]
```diff
@@ -441,10 +453,16 @@ impl<Descriptor: SocketDescriptor, CM: Deref, RM: Deref, L: Deref> PeerManager<D
 			ephemeral_key_midstate,
 			peer_counter_low: AtomicUsize::new(0),
 			peer_counter_high: AtomicUsize::new(0),
+			peer_connectivity_updater: None,
```
Comment (jkczyz) [inline lightning/src/ln/peer_handler.rs:456]: Then name this perhaps `connectivity_listener` as the actual value is listening for updates not notifying of updates (which is what `PeerManager` is doing).

File: lightning/src/ln/peer_handler.rs [hunk 4]
```diff
@@ -236,6 +236,14 @@ impl error::Error for PeerHandleError {
 	}
 }
 
+/// Trait for subscribing to changes of a peer manager's peer set
+pub trait PeerConnectivityUpdater {
+	/// Called when a new peer connects (once the handshake is finished)
+	fn peer_connected(&self, their_node_id: &PublicKey);
+	/// Called when a peer disconnects (for any reason) or is disconnected (watch out for reëntrancy bugs)
+	fn peer_disconnected(&self, their_node_id: &PublicKey);
```
Comment (jkczyz) [inline lightning/src/ln/peer_handler.rs:244]: nit: drop "their"

**Review by jkczyz — COMMENTED** [review]
Did you consider adding `Event`s for these? Curious as to the tradeoffs of either approach.

**Comment by TheBlueMatt** [comment]
Closing as outdated. Can reopen when/if this gets rebased.

**Pull request closed without merging**